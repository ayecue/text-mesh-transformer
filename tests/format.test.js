const { transform, Tag, TagRecordOpen, TagRecordClose } = require('../dist/index.js');

const testCallback = (openTag, content) => {
  if (openTag.isWrappedBy(Tag.NoParse)) {
    return `${openTag.raw}${content}${openTag.next.raw}`
  }

  switch (openTag.type) {
    case Tag.Color: {
      return `[myColor=${openTag.attributes.value}]${content}[/myColor]`;
    }
    case Tag.Mark: {
      return `[myMark=${openTag.attributes.value}]${content}[/myMark]`;
    }
    case Tag.Underline:
      return `[underline]${content}[/underline]`;
    case Tag.Italic:
      return `[italic]${content}[/italic]`;
    case Tag.Bold:
      return `[bold]${content}[/bold]`;
    case Tag.Font:
      return `[font="${openTag.attributes.value}"]${content}[/font]`;
    case Tag.Sprite:
      return `[sprite="${openTag.attributes.value}", "${openTag.attributes.foo}"]${content}[/sprite]`;
  }

  return content;
};

describe('transform', function () {
  test('simple', function () {
    const result = transform(`
    was <color=red>test <color=blue><u><b>wd</b></u></color> xad
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('tag with two attributes', function () {
    const result = transform(`
    was <sprite=0 foo=somename>test xad
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('without close tags', function () {
    const result = transform(`
    was <color=red>test <color=blue><u><b>wd xad
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('unknown tags', function () {
    const result = transform(`
    was <color=red>test <unknown>wd xad
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('invalid closing of tag', function () {
    const result = transform(`
    was <color=red>test <b>wd xad</color>
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('two top level tags', function () {
    const result = transform(`
    was <color=red>test</color> <color=blue><b>wd xad
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('two top level tags with inner styles', function () {
    const result = transform(`
    was <color=red>te<u>st</u></color> <color=blue><b>wd xad
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('style without value', function () {
    const result = transform(`
    was <link>test</link>
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('style with value in quotes', function () {
    const result = transform(`
    was <font=" my test  value">test</font>
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('color shorthand', function () {
    const result = transform(`
    was <#333666>test</color>
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('context', function () {
    const result = transform(`
    was <noparse><#333666>test</color>
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('multiline', function () {
    const result = transform(`
    <color=red>
    title - text

    -a --abc - <b>description</b>
    -a --abc - <b>description</b>
    -a --abc - <b>description</b>
    -a --abc - <b>description</b>
    </color>
    `, testCallback);

    expect(result).toMatchSnapshot();
  });

  test('multiline 2', function() {
    const result = transform(`<b><u>tedit</u></b>
    <color=yellow>1 | 	</color><mark=#ff00aa>$
    2 | 	<color=#FFF>test</color> = <color=#E11584>function</color>()
    3 | 		<color=#FFF>myTest</color> = <color=#F9D71C>"hello world"</color>
    4 | 		<color=#009dc4>print</color>(<color=#FFF>myTest</color>)
    5 | 	<color=#E11584>end function</color>
    6 | 	
    7 | 	<color=#FFF>test</color>
    8 | 
    `, testCallback);

    expect(result).toMatchSnapshot();
  })

  test('newline', function() {
    const result = transform([
      "<b>Title: ",
      "   <b>description",
      "       text: text",
      "       text: text",
      "",
      "   <color=red>bet (amount) (slotNumber)",
      "       ex:  bet 100 7",
      "       ex:  bet 100 0",
      "       ex:  bet 100 00"
    ].join('\n'), testCallback);

    expect(result).toMatchSnapshot();
  });
});
