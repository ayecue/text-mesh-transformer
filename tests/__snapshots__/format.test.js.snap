// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform color shorthand 1`] = `
"
    was [myColor=#333666]test[/myColor]
    "
`;

exports[`transform context 1`] = `
"
    was <#333666>test</color>
    "
`;

exports[`transform invalid closing of tag 1`] = `
"
    was [myColor=red]test [bold]wd xad</color>[/bold][/myColor]
    "
`;

exports[`transform multiline 1`] = `
"
    [myColor=red][/myColor]
    title - text

    -a --abc - [bold]description[/bold]
    -a --abc - [bold]description[/bold]
    -a --abc - [bold]description[/bold]
    -a --abc - [bold]description[/bold]
    </color>
    "
`;

exports[`transform multiline 2 1`] = `
"[bold][underline]tedit[/underline][/bold]
    [myColor=yellow]1 | 	[/myColor][myMark=#ff00aa]$[/myMark]
    2 | 	[myColor=#FFF]test[/myColor] = [myColor=#E11584]function[/myColor]()
    3 | 		[myColor=#FFF]myTest[/myColor] = [myColor=#F9D71C]\\"hello world\\"[/myColor]
    4 | 		[myColor=#009dc4]print[/myColor]([myColor=#FFF]myTest[/myColor])
    5 | 	[myColor=#E11584]end function[/myColor]
    6 | 	
    7 | 	[myColor=#FFF]test[/myColor]
    8 | 
    "
`;

exports[`transform newline 1`] = `
"[bold]Title: [/bold]
   [bold]description[/bold]
       text: text
       text: text

   [myColor=red]bet (amount) (slotNumber)[/myColor]
       ex:  bet 100 7
       ex:  bet 100 0
       ex:  bet 100 00"
`;

exports[`transform simple 1`] = `
"
    was [myColor=red]test [myColor=blue][underline][bold]wd[/bold][/underline][/myColor] xad[/myColor]
    "
`;

exports[`transform style with value in quotes 1`] = `
"
    was [font=\\" my test  value\\"]test[/font]
    "
`;

exports[`transform style without value 1`] = `
"
    was test
    "
`;

exports[`transform tag with two attributes 1`] = `
"
    was [sprite=\\"0\\", \\"somename\\"][/sprite]test xad
    "
`;

exports[`transform two top level tags 1`] = `
"
    was [myColor=red]test[/myColor] [myColor=blue][bold]wd xad[/bold][/myColor]
    "
`;

exports[`transform two top level tags with inner styles 1`] = `
"
    was [myColor=red]te[underline]st[/underline][/myColor] [myColor=blue][bold]wd xad[/bold][/myColor]
    "
`;

exports[`transform unknown tags 1`] = `
"
    was [myColor=red]test <unknown>wd xad[/myColor]
    "
`;

exports[`transform without close tags 1`] = `
"
    was [myColor=red]test [myColor=blue][underline][bold]wd xad[/bold][/underline][/myColor][/myColor]
    "
`;
